File: json/json-annotation-processor/src/test/java/ru/tinkoff/kora/json/annotation/processor/GenericsTest.java
package ru.tinkoff.kora.json.annotation.processor;

import org.junit.jupiter.api.Test;
import org.mockito.Mockito;
import ru.tinkoff.kora.json.common.JsonReader;
import ru.tinkoff.kora.json.common.JsonWriter;
import ru.tinkoff.kora.kora.app.annotation.processor.KoraAppProcessor;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

@SuppressWarnings("unchecked")
public class GenericsTest extends AbstractJsonAnnotationProcessorTest {
    @Test
    public void testRecordWithTypeParameter() {
        compile("""
            @Json
            public record TestRecord <T>(T value, java.util.List<T> values) {
            }
            """);
        var r1 = Mockito.mock(JsonReader.class);
        var r2 = Mockito.mock(JsonReader.class);
        var w1 = Mockito.mock(JsonWriter.class);
        var w2 = Mockito.mock(JsonWriter.class);
        var m1 = mapper("TestRecord", List.of(r1, r2), List.of(w1, w2));
    }

    @Test
    public void testRecordWithTypeParametersFromExtension() throws IOException, ClassNotFoundException {
        compile(List.of(new JsonAnnotationProcessor(), new KoraAppProcessor()), """
                @Json
                public record TestRecord <T>(T value, java.util.List<T> values) {
                }
                """,
            """
                @KoraApp
                public interface TestApp extends ru.tinkoff.kora.json.common.JsonCommonModule {
                    @Root
                    default String root1(ru.tinkoff.kora.json.common.JsonWriter<TestRecord<Integer>> w, ru.tinkoff.kora.json.common.JsonReader<TestRecord<Integer>> r) { return ""; }
                    @Root
                    default String root2(ru.tinkoff.kora.json.common.JsonWriter<TestRecord<String>> w, ru.tinkoff.kora.json.common.JsonReader<TestRecord<String>> r) { return ""; }
                }
                    """);
        compileResult.assertSuccess();
        var graph = loadGraph("TestApp");
        var reader = (JsonReader) graph.findByType(compileResult.loadClass("$TestRecordJsonReader"));
        var writer = (JsonWriter) graph.findByType(compileResult.loadClass("$TestRecordJsonWriter"));

        var o = newObject("TestRecord", 42, List.of(42, 43));
        var json = "{\"value\":42,\"values\":[42,43]}";

        assertThat(reader.read(json.getBytes(StandardCharsets.UTF_8))).isEqualTo(o);
        assertThat(writer.toByteArray(o)).asString(StandardCharsets.UTF_8).isEqualTo(json);
    }

    @Test
    public void testRecordWithTypeParametersFromExtensionNoAnnotations() throws IOException, ClassNotFoundException {
        compile(List.of(new JsonAnnotationProcessor(), new KoraAppProcessor()), """
                public record TestRecord <T>(T value, java.util.List<T> values) {
                }
                """,
            """
                @KoraApp
                public interface TestApp extends ru.tinkoff.kora.json.common.JsonCommonModule {
                    @Root
                    default String root1(ru.tinkoff.kora.json.common.JsonWriter<TestRecord<Integer>> w, ru.tinkoff.kora.json.common.JsonReader<TestRecord<Integer>> r) { return ""; }
                    @Root
                    default String root2(ru.tinkoff.kora.json.common.JsonWriter<TestRecord<String>> w, ru.tinkoff.kora.json.common.JsonReader<TestRecord<String>> r) { return ""; }
                }
                    """);
        compileResult.assertSuccess();
        var graph = loadGraph("TestApp");
        var reader = (JsonReader) graph.findByType(compileResult.loadClass("$TestRecordJsonReader"));
        var writer = (JsonWriter) graph.findByType(compileResult.loadClass("$TestRecordJsonWriter"));

        var o = newObject("TestRecord", 42, List.of(42, 43));
        var json = "{\"value\":42,\"values\":[42,43]}";

        assertThat(reader.read(json.getBytes(StandardCharsets.UTF_8))).isEqualTo(o);
        assertThat(writer.toByteArray(o)).asString(StandardCharsets.UTF_8).isEqualTo(json);
    }
}

File: json/json-symbol-processor/src/test/kotlin/ru/tinkoff/kora/json/ksp/GenericsTest.kt
package ru.tinkoff.kora.json.ksp

import org.junit.jupiter.api.Test
import ru.tinkoff.kora.json.common.JsonReader
import ru.tinkoff.kora.json.common.JsonWriter
import ru.tinkoff.kora.kora.app.ksp.KoraAppProcessorProvider
import ru.tinkoff.kora.ksp.common.GraphUtil.toGraph

class GenericsTest : AbstractJsonSymbolProcessorTest() {
    @Test
    fun testGenericJsonReaderExtension() {
        compile(
            listOf(KoraAppProcessorProvider()),
            """
            data class TestClass <T> (val value:T)             
            """.trimIndent(),
            """
                @KoraApp
                interface TestApp : ru.tinkoff.kora.json.common.JsonCommonModule {
                  @Root
                  fun root(w1: ru.tinkoff.kora.json.common.JsonReader<TestClass<String>>, w2: ru.tinkoff.kora.json.common.JsonReader<TestClass<Int>>) = ""
                }
            """.trimIndent()
        )
        val graph = loadClass("TestAppGraph").toGraph()
        val reader = graph.findAllByType(readerClass("TestClass")) as List<JsonReader<Any?>>

        reader[0].assertRead("{\"value\":\"test\"}", new("TestClass", "test"))
        reader[1].assertRead("{\"value\":42}", new("TestClass", 42))
    }

    @Test
    fun testGenericJsonWriterExtension() {
        compile(
            listOf(KoraAppProcessorProvider()),
            """
            data class TestClass <T> (val value:T)             
            """.trimIndent(),
            """
                @KoraApp
                interface TestApp : ru.tinkoff.kora.json.common.JsonCommonModule {
                  @Root
                  fun root(w1: ru.tinkoff.kora.json.common.JsonWriter<TestClass<String>>, w2: ru.tinkoff.kora.json.common.JsonWriter<TestClass<Int>>) = ""
                }
            """.trimIndent()
        )
        val graph = loadClass("TestAppGraph").toGraph()
        val writer = graph.findAllByType(writerClass("TestClass")) as List<JsonWriter<Any?>>

        writer[0].assertWrite(new("TestClass", "test"), "{\"value\":\"test\"}")
        writer[1].assertWrite(new("TestClass", 42), "{\"value\":42}")
    }

    @Test
    fun testGenericJsonReaderExtensionWithAnnotation() {
        compile(
            listOf(KoraAppProcessorProvider(), JsonSymbolProcessorProvider()),
            """
            @ru.tinkoff.kora.json.common.annotation.Json
            data class TestClass <T> (val value:T)             
            """.trimIndent(),
            """
                @KoraApp
                interface TestApp : ru.tinkoff.kora.json.common.JsonCommonModule {
                  @Root
                  fun root(w1: ru.tinkoff.kora.json.common.JsonReader<TestClass<String>>, w2: ru.tinkoff.kora.json.common.JsonReader<TestClass<Int>>) = ""
                }
            """.trimIndent()
        )
        val graph = loadClass("TestAppGraph").toGraph()
        val reader = graph.findAllByType(readerClass("TestClass")) as List<JsonReader<Any?>>

        reader[0].assertRead("{\"value\":\"test\"}", new("TestClass", "test"))
        reader[1].assertRead("{\"value\":42}", new("TestClass", 42))
    }

    @Test
    fun testGenericJsonWriterExtensionWithAnnotation() {
        compile(
            listOf(KoraAppProcessorProvider(), JsonSymbolProcessorProvider()),
            """
            @ru.tinkoff.kora.json.common.annotation.Json
            data class TestClass <T> (val value:T)             
            """.trimIndent(),
            """
                @KoraApp
                interface TestApp : ru.tinkoff.kora.json.common.JsonCommonModule {
                  @Root
                  fun root(w1: ru.tinkoff.kora.json.common.JsonWriter<TestClass<String>>, w2: ru.tinkoff.kora.json.common.JsonWriter<TestClass<Int>>) = ""
                }
            """.trimIndent()
        )
        val graph = loadClass("TestAppGraph").toGraph()
        val writer = graph.findAllByType(writerClass("TestClass")) as List<JsonWriter<Any?>>

        writer[0].assertWrite(new("TestClass", "test"), "{\"value\":\"test\"}")
        writer[1].assertWrite(new("TestClass", 42), "{\"value\":42}")
    }

}
